From 8a2befbade9f94f40bf86eb7907ab1e5622411ba Mon Sep 17 00:00:00 2001
From: Juan Lang <juan.lang@ericsson.com>
Date: Thu, 10 Sep 2009 14:14:20 -0700
Subject: [PATCH 60/67] Support connecting to canonical names

---
 net/namestack/af_name.c |   20 ++++++++++++++++++++
 1 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/net/namestack/af_name.c b/net/namestack/af_name.c
index 919b875..8e2287c 100644
--- a/net/namestack/af_name.c
+++ b/net/namestack/af_name.c
@@ -1337,6 +1337,26 @@ static int name_stream_connect(struct socket *sock, struct sockaddr *uaddr,
 			err = name_stream_connect_to_v6_address(sk, sizeof(in6),
 								(const u_char *)&in6);
 		}
+		else if (name_find_v4_canonical_suffix(
+			name->dname.sname_addr.name) != NULL) {
+			__be32 v4;
+
+			err = name_parse_canonical_v4(
+				name->dname.sname_addr.name, &v4);
+			if (!err)
+				err = name_stream_connect_to_v4_address(sk,
+					sizeof(v4), (const u_char *)&v4);
+		}
+		else if (name_find_v6_canonical_suffix(
+			name->dname.sname_addr.name) != NULL) {
+			struct in6_addr in6;
+
+			err = name_parse_canonical_v6(
+				name->dname.sname_addr.name, &in6);
+			if (!err)
+				err = name_stream_connect_to_v6_address(sk,
+					sizeof(in6), in6.s6_addr);
+		}
 		else
 			err = name_send_query(sname->sname_addr.name,
 					      name_stream_query_resolve, sock);
-- 
1.6.3.2

