From 1bc1294b96074ae563583079e9c3e71404ca9e5f Mon Sep 17 00:00:00 2001
From: Juan Lang <juan.lang@ericsson.com>
Date: Thu, 10 Sep 2009 13:51:20 -0700
Subject: [PATCH] Implement sendmsg

---
 af_name.c |   20 +++++++++++++++++++-
 1 files changed, 19 insertions(+), 1 deletions(-)

diff --git a/af_name.c b/af_name.c
index 2c1593c..a3f3022 100644
--- a/af_name.c
+++ b/af_name.c
@@ -944,6 +944,24 @@ out:
 	return err;
 }
 
+static int name_stream_sendmsg(struct kiocb *iocb, struct socket *sock,
+			       struct msghdr *msg, size_t len)
+{
+	struct sock *sk = sock->sk;
+	struct name_stream_sock *name = name_stream_sk(sk);
+	struct socket *connected_sock;
+
+	if (sock->state != SS_CONNECTED)
+		return -ENOTCONN;
+	if (name->ipv6_sock)
+		connected_sock = name->ipv6_sock;
+	else if (name->ipv4_sock)
+		connected_sock = name->ipv4_sock;
+	else
+		return -ENOTCONN;
+	return connected_sock->ops->sendmsg(iocb, connected_sock, msg, len);
+}
+
 static const struct proto_ops name_stream_ops = {
 	.family = PF_NAME,
 	.owner = THIS_MODULE,
@@ -962,7 +980,7 @@ static const struct proto_ops name_stream_ops = {
 	.shutdown = sock_no_shutdown,
 	.setsockopt = sock_no_setsockopt,
 	.getsockopt = sock_no_getsockopt,
-	.sendmsg = sock_no_sendmsg,
+	.sendmsg = name_stream_sendmsg,
 	.recvmsg = sock_no_recvmsg,
 	.mmap = sock_no_mmap,
 	.sendpage = sock_no_sendpage,
-- 
1.6.0.4

